src/java.base/share/classes/java/math/BigDecimal.java

```diff
@@ -3732,31 +3746,69 @@ public byte byteValueExact() {
      * @jls 5.1.3 Narrowing Primitive Conversion
      */
     @Override
-    public float floatValue(){
-        if(intCompact != INFLATED) {
+    public float floatValue() {
+        /* For details, see the extensive comments in doubleValue(). */
+        if (intCompact != INFLATED) {
+            float v = intCompact;
             if (scale == 0) {
-                return (float)intCompact;
-            } else {
-                /*
-                 * If both intCompact and the scale can be exactly
-                 * represented as float values, perform a single float
-                 * multiply or divide to compute the (properly
-                 * rounded) result.
-                 */
-                if (Math.abs(intCompact) < 1L<<22 ) {
-                    // Don't have too guard against
-                    // Math.abs(MIN_VALUE) because of outer check
-                    // against INFLATED.
-                    if (scale > 0 && scale < FLOAT_10_POW.length) {
-                        return (float)intCompact / FLOAT_10_POW[scale];
-                    } else if (scale < 0 && scale > -FLOAT_10_POW.length) {
-                        return (float)intCompact * FLOAT_10_POW[-scale];
-                    }
+                return v;
+            }
+            /*
+             * The discussion for the double case also applies here. That is,
+             * the following test is precise for all long values except for
+             * Long.MAX_VALUE but the result is correct nevertheless.
+             */
+            if ((long) v == intCompact) {
+                if (0 < scale && scale < FLOAT_10_POW.length) {
+                    return v / FLOAT_10_POW[scale];
+                }
+                if (0 > scale && scale > -FLOAT_10_POW.length) {
+                    return v * FLOAT_10_POW[-scale];
                 }
             }
         }
-        // Somewhat inefficient, but guaranteed to work.
-        return Float.parseFloat(this.toString());
+        return fullFloatValue();
+    }
+
+    private float fullFloatValue() {
+        if (intCompact == 0) {
+            return 0.0f;
+        }
+        BigInteger d = unscaledValue().abs();
```

Comment:
I'd prefer a name other than "d" be used for the BigInteger significand's magnitude.

Reply:
> I have no objections as long as it is a 1 letter name that does not conflict with others. Have you a preferred one?

I usually use "bd" or "bi" for such BigDecima/BigInteger variables myself and use "d" for double; "v" or "w" have integer-ish feel to me so perhaps one of those?